src/LanguageServer/Protocol/Handler/SourceGenerators/SourceGeneratorRefreshQueue.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SourceGenerators;
+
+internal class SourceGeneratorRefreshQueue :
+        IOnInitialized,
+        ILspService,
+        IDisposable
+{
+    private const string RefreshSourceGeneratedDocumentName = "workspace/refreshSourceGeneratedDocument";
+
+    private readonly IAsynchronousOperationListener _asyncListener;
+    private readonly CancellationTokenSource _disposalTokenSource = new();
+    private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService;
+    private readonly LspWorkspaceManager _lspWorkspaceManager;
+    private readonly IClientLanguageServerManager _notificationManager;
+    private readonly AsyncBatchingWorkQueue _refreshQueue;
+
+    public SourceGeneratorRefreshQueue(
+                IAsynchronousOperationListenerProvider asynchronousOperationListenerProvider,
+                LspWorkspaceRegistrationService lspWorkspaceRegistrationService,
+                LspWorkspaceManager lspWorkspaceManager,
+                IClientLanguageServerManager notificationManager)
+    {
+        _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
+        _lspWorkspaceManager = lspWorkspaceManager;
+        _notificationManager = notificationManager;
+        _asyncListener = asynchronousOperationListenerProvider.GetListener(FeatureAttribute.SourceGenerators);
+
+        // Batch up workspace notifications so that we only send a notification to refresh source generated files
+        // every 2 seconds - long enough to avoid spamming the client with notifications, but short enough to refresh
+        // the source generated files relatively frequently.
+        _refreshQueue = _refreshQueue = new AsyncBatchingWorkQueue(
+            delay: TimeSpan.FromMilliseconds(2000),
```

Comment:
nit: we have 1.5s for this (as a TaggerDelay defined somewhere) as this has been empirically tested as the delay that bothers people least, while backing off enough to get a lot of value.

Reply:
DelayTimeSpan.Idle.